### PR TITLE
lecture 18: Deep Generative Models (Part 2) -- v1

### PR DESCRIPTION
**Pull request by RifleZhang**
Link preview: https://riflezhang.github.io/pgm-spring-2019/notes/lecture-18/
I submitted the first version of lecture 18 scribing.
Thanks for revision!

### CODE REVIEW

**Comment by RLAgent** [comment]
Thanks for the first draft of notes.

The content looks good, but could you format the scribe notes in paragraph form so that it reads more like a blog post, rather than bullet points from lecture slides.

Here's an example that I liked: https://sailinglab.github.io/pgm-spring-2019/notes/lecture-13/

File: _posts/2019-03-25-lecture-18.md [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+
+### Vanilla GAN Review
+In GAN setting, we have a generative model and a discriminator model. The generative model trys to generate new data given a prior, 
+while the discriminative model distinguishes the difference between the generated data and the true data.
+
+
+The **generative** part of GAN model maps a random noise to a generated data, i.e. $\bm{x} = G_\theta(\bm{z}), \bm{z} \sim p(\bm{z})$, where $z$ is a random noise and $x$ is the data space.
+It defines a implicit distribution over $x: p_{g_\theta}(x)$, and uses a stochastic process to simulate data $x$. It is intractable to evaluate the likelihood of generated data.
+
+The **descriminative** part of GAN tells apart the true data distibution from the generated distribution. Specifically, the descrimination $D_\phi (x)$ outputs a probability of $x$ being the true data.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]: Spelling: "discriminative"

```suggestion
The **discriminative** part of GAN tells apart the true data distribution from the generated distribution. Specifically, the discriminator $D_\phi (x)$ outputs a probability of $x$ being the true data.
```

File: _posts/2019-03-25-lecture-18.md [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+editors:
+  - name: Lisa Lee
+    url: "#"  # optional URL to the editor's homepage
+
+---
+
+<!-- Ruohong -->
+## GANs and their variations
+### GAN Progress
+From 2014 to 2018, GAN can generate more high quality and more realistic images.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
Between 2014 and 2018, GAN has progressively been able to generate more high quality and more realistic images.
```

File: _posts/2019-03-25-lecture-18.md [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+<figure>
+<img src="{{ '/assets/img/notes/lecture-18/gan_progress.jpeg' | relative_url }}" />
+<figcaption>
+The changing of quality of image generation using GAN. 
+Figure courtesy: Ian Goodfellow
+</figcaption>
+</figure>
+
+### Vanilla GAN Review
+In GAN setting, we have a generative model and a discriminator model. The generative model trys to generate new data given a prior,
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
In the GAN setting, we have a generative model and a discriminator model. The generative model tries to generate new data given a prior, 
```

File: _posts/2019-03-25-lecture-18.md [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+</figcaption>
+</figure>
+
+### Vanilla GAN Review
+In GAN setting, we have a generative model and a discriminator model. The generative model trys to generate new data given a prior, 
+while the discriminative model distinguishes the difference between the generated data and the true data.
+
+
+The **generative** part of GAN model maps a random noise to a generated data, i.e. $\bm{x} = G_\theta(\bm{z}), \bm{z} \sim p(\bm{z})$, where $z$ is a random noise and $x$ is the data space.
+It defines a implicit distribution over $x: p_{g_\theta}(x)$, and uses a stochastic process to simulate data $x$. It is intractable to evaluate the likelihood of generated data.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
It defines an implicit distribution over $x: p_{g_\theta}(x)$, and uses a stochastic process to simulate data $x$. It is intractable to evaluate the likelihood of the generated data.
```

File: _posts/2019-03-25-lecture-18.md [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+<img src="{{'/assets/img/notes/lecture-18/gan_pipeline.png' | relative_url }}" />
+<figcaption>
+GAN pipeline on MNIST Data
+</figcaption>
+</figure>
+
+### Wasserstein GAN(WGAN)
+In the setting of vanilla GAN, if the data and model's manifold are different, there can be negligible intersection between the two in practice. In this case, the loss function and gradients may not be continous or well-behaved. For example, for some $x$ such that $P_g(x)=0$ but $P_d(x) > 0$, KL divergence is infinite.
+
+In order to solve the problem, **WGAN** <d-cite key="wgan"></d-cite> was proposed, in which a new objective was used by changing the distance function to Wasserstein Distance(Earth Mover). It measures the minimum transportation cost for making one pile of dirt in the shape of one probability distribution to the shape of the other distribution.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
In order to solve the problem, **WGAN** <d-cite key="wgan"></d-cite> was proposed, in which a new objective was used by changing the distance function to Wasserstein Distance (Earth Mover). It measures the minimum transportation cost for making one pile of dirt in the shape of one probability distribution to the shape of the other distribution.
```

File: _posts/2019-03-25-lecture-18.md [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+
+The hard part of learning objective is to ensure that $\|D\|_{L} \leq K$ K-Lipschitz continuous. In the paper, it uses gradient-clipping to ensure $D$ has the Lipschitz continuity, which was acknowledged by the author that it may be a terrible way to do so, but it is simple, yet still effective, and performs better the vanilla GAN.
+<d-math block>
+W (p_d, p_g) = {1 \over K} \cdot sup_{\|D\|_{L} \leq K} E_{x \sim p_d}[D(x)] - E_{x \sim p_g}[D(x)]
+</d-math>
+
+
+During the training, we compare the GAN discriminator with the WGAN critic when learning to differentiate two Gaussians. As we can see, the traditional GAN
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
During training, we compare the GAN discriminator with the WGAN critic when learning to differentiate between two Gaussians. As we can see, the traditional GAN
```

File: _posts/2019-03-25-lecture-18.md [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+<figure>
+<img src="{{'/assets/img/notes/lecture-18/wgan_training.png' | relative_url }}" />
+<figcaption>
+WGAN Training
+</figcaption>
+</figure>
+
+### Progressive GAN
+Although GAN can produce sharp images, in most of the cases those images are in small resolutions with limited variation. Directly generating large image is impractable because the training becomes unstable and the model is hard to train. Due to those problems, the **Progressive GAN** <d-cite key="pgan"></d-cite> was proposed to improve the training method for large images.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
Although GAN can produce sharp images, in most cases, the images are in small resolutions with limited variation. Directly generating large image is impractical due to training instability. Due to those problems, the **Progressive GAN** <d-cite key="pgan"></d-cite> was proposed to improve the training method for large images. 
```

**Review by RLAgent — CHANGES_REQUESTED** [review]
Thanks for the updated notes!

I addressed some grammatical issues here: https://github.com/sailinglab/pgm-spring-2019/pull/21/files#diff-7130e1985c0ff084366496441af74cb8

It looks like "Integrating domain knowledge in DL" is still in bullet points, so could you ping me when this is changed.

**Comment by alshedivat** [comment]
@lileee, @RifleZhang, just checking in to see if these notes are ready to be merged. Thanks.

Btw, the first image in the notes is pretty low resolution (which kind of defeats the point :)). How about using [this one](https://pbs.twimg.com/media/Dw6ZIOlX4AMKL9J.jpg) instead?

**Comment by alshedivat** [comment]
After a quick look, seems like "Learning with constraints" section needs another pass (basically, convert bullet points to text). Minor: the reference under the GLOW samples image is plain text -- could you make it `<d-cite></d-cite>` type of reference?

**Comment by RifleZhang** [comment]
Will address the problems soon.
Thanks!

File: _posts/2019-03-25-lecture-18.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+In the M step, we plug in $q^*$ and optimize w.r.t. $\theta$ with:
+<d-math block>
+\min_\theta KL(q(x)\|p_\theta(x))=\min_\theta \mathcal{L}(\theta) - E_{q^*}[\log p_\theta(x)] 
+</d-math> 
+
+#### Learnable constraints
+In previous formulation, constraint $f(x)$ has to be fully-specified a priori and is fixed throughout the learning. It would be desirable or necessary to enable learnable constraint, so that people can specify the know constraints, while leaving the unknown parts learned by the model. We would like to introduce learnable parameters into the constraint function, such that we rewrite $f(x)$ as $f_\phi(x)$ where $\phi$ are learnable parameters. Learning $\phi$ is not straightforward, if we optimize $\phi$ in M-step, i.e. 
+$$\max_\phi E_{x\sim q(x)}[f_\phi(x)]$$, such objective could be problematic due to the poor state of generative parameter $\theta$ at the initial stage. Instead, we can treat learning of constraint as an extrinsic reward motived by Reinforcement Learning, and get a joint learning algorithm for $\theta$ and $\phi$. 
+
+## Learning with Constraint
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
## Learning with Constraints
```

File: _posts/2019-03-25-lecture-18.md [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+
+#### knowledge constraint
+In order to evaluate the structural difference between the generated image and the target image, we introduce a constraint function $f_\theta(x)$ that encourages each of the body parts (e.g., head, legs) of the generated image to match the respective part of the target image. Specifically, the constraint $f_\theta(x)$ includes a human parsing module that classifies each pixel of a person image into possible body parts, and then evaluates cross entropies of the per-pixel part distributions between the generated and target images. The average negative cross entropy serves as the constraint score, and higher $f_\theta(x)$ values indicates a better generated $x$ in terms of pariticular domain knowledge.
+
+#### learning objectives
+In the previous sections, the propsed model has two parts: a deep generative model and a knowledge constraint model. In the optimization step, we should have a regular generative objective $\mathcal{L}(\theta)$, and a constraint regularization expectation $E_{p_\theta}[f(x)]$. We put them together, the objective is simply:
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
In the previous sections, the proposed model has two parts: a deep generative model and a knowledge constraint model. In the optimization step, we should have a regular generative objective $\mathcal{L}(\theta)$, and a constraint regularization expectation $E_{p_\theta}[f(x)]$. Putting them together, the objective is simply:
```

File: _posts/2019-03-25-lecture-18.md [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+<d-math block>
+\min_\theta \mathcal{L}(\theta) - \alpha E_{p_\theta}[f(x)]
+</d-math> 
+
+In general,$E_{p_\theta}[f(x)]$ is hard to optimize unless $p_\theta$ is a GAN-like implicit generative model or an explicit distribution that can be efficiently reparametrized.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
In general, $E_{p_\theta}[f(x)]$ is hard to optimize unless $p_\theta$ is a GAN-like implicit generative model or an explicit distribution that can be efficiently reparametrized.
```

File: _posts/2019-03-25-lecture-18.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+</d-math>
+Example:
+<figure>
+<img src="{{'/assets/img/notes/lecture-18/gan_pipeline.png' | relative_url }}" />
+<figcaption>
+GAN pipeline on MNIST Data
+</figcaption>
+</figure>
+
+### Wasserstein GAN(WGAN)
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
### Wasserstein GAN (WGAN)
```

File: _posts/2019-03-25-lecture-18.md [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+</d-math> 
+
+In general,$E_{p_\theta}[f(x)]$ is hard to optimize unless $p_\theta$ is a GAN-like implicit generative model or an explicit distribution that can be efficiently reparametrized.
+For other models such as the large set of non-reparametrizable explicit distributions, the gradient $\nabla E_{p_\theta}[f_\theta(x)]$ is usually computed with the log-derivative trick and can suffer from high variance. For broad applicability and efficient optimization, PR instead imposes the constraint on an auxiliary variational distribution $q$, and introduces a KL divergence term:
+<d-math block>
+\mathcal{L}(\theta,q)=KL(q(x) \| p_\theta(x)) - \alpha E_q[f(x)] 
+</d-math> 
+as proxy of $- E_{p_\theta}[f(x)]$
+
+The optimization problem could be solve by EM algorithm.
```
Comment (RLAgent) [inline _posts/2019-03-25-lecture-18.md]:
```suggestion
The optimization problem can be solve by EM algorithm:
```

**Review by RLAgent — CHANGES_REQUESTED** [review]
The notes look good, thanks!

I made minor grammatical error comments here: https://github.com/sailinglab/pgm-spring-2019/pull/21/files#diff-7130e1985c0ff084366496441af74cb8

Could you address them, then we can merge the notes to master. Thanks!

**Comment by RifleZhang** [comment]
Dear Lisa,

The changes are merged. Is there anything else to address before merging?
Thanks!

Best,
Ruohong

**Pull request merged**